### PR TITLE
[WIP] Transformer Chainer

### PR DESCRIPTION
**Pull request by Fhrozen**
The initial commit to support transformer in chainer. This is a extension of #465 which will be closed.

Some detail:
- It is most compatible with pytorch code, but some functions are required to be modified.
- Added test but not for chainer (currently)
- Because of the implementation is no longer recurrent, the use of `loss.unchain_backward()` seems unnecessary in the training of the transformer. It also seems to affect the training time (Need more test). If this is removed during training the transformer, then the visualization of the full network becomes possible. Let me know if this is required.
- The model only trains in att mode, and decodes with greedy search.  Additional details are WIP.

I am testing it to submit the initial results. The log says that the training will be finishing in a 33 hrs (quite larger than any additional test I did).

### CODE REVIEW

**Comment by sw005320** [comment]
There seem to have some bash script issues in https://travis-ci.org/espnet/espnet/jobs/510541041#L1341-L1350
Please fix them.

@gtache, the bash script check (https://github.com/espnet/espnet/blob/master/.travis.yml#L77) does not seem to exit with the appropriate error code. Could you check it?

**Comment by sw005320** [comment]
We first merge this, and move to the pytorch backend.

File: egs/wsj/asr2/run.sh [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+mkdir -p ${expdir}
+
+if [ ${stage} -le 4 ] && [ ${stop_stage} -ge 4 ]; then
+    echo "stage 4: Network Training"
+
+    ${cuda_cmd} --gpu ${ngpu} ${expdir}/train.log \
+        asr_train.py \
+        --accum-grad ${accum_grad} \
+        --ngpu ${ngpu} \
+        --backend ${backend} \
```
Comment (ShigekiKarita) [inline egs/wsj/asr2/run.sh]: I think this `--backend ${backend}` and `--model-module "espnet.nets.${backend}_backend.asr_transformer" ` are redundant. how about identifing the backend in `asr_train.py` like
```python
module = load_module(args.model_module)
args.backend = "chainer" if isinstance(module.E2E, chainer.Chain) else "pytorch"
```

File: egs/wsj/asr2/run.sh [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+        --dropout-rate ${dropout} \
+        --maxlen-in ${maxlen_in} \
+        --maxlen-out ${maxlen_out} \
+        --opt ${opt} \
+        --grad-clip ${grad_clip} \
+        --sampling-probability ${samp_prob} \
+        --epochs ${epochs} \
+        --sortagrad ${sortagrad} \
+        --lsm-weight ${lsm_weight} \
+        --model-module "espnet.nets.${backend}_backend.asr_transformer" \
```
Comment (ShigekiKarita) [inline egs/wsj/asr2/run.sh]: It is not user-friendly to assume implicitly that `E2E` class is defined in this module. How about specifying explicitly `espnet.nets.${backend}_backend.asr_transformer.E2E`  instead?

File: egs/wsj/asr2/run.sh [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+verbose=0      # verbose option
+resume=        # Resume the training from snapshot
+seed=1
+
+# feature configuration
+do_delta=false
+
+# network architecture
+ninit=pytorch
+ntype=transformer
```
Comment (ShigekiKarita) [inline egs/wsj/asr2/run.sh]: I think $ntype is only used in $expdir and it does not change `asr_train.py` behavior. How about new cmd arg `mode_lmodule="espnet.nets.${backend}_backend.asr_transformer.E2E"` instead to indicate the exact model?

File: espnet/asr/chainer_backend/asr.py [hunk 4]
```diff
@@ -104,34 +108,42 @@ def update_core(self):
         logging.info('grad norm={}'.format(grad_norm))
         if math.isnan(grad_norm):
             logging.warning('grad norm is nan. Do not update model.')
-        else:
+        elif self.count % self.accum_grad == 0:
             optimizer.update()
+            self.clean = True
```
Comment (ShigekiKarita) [inline espnet/asr/chainer_backend/asr.py]: why don't you clean up gradients immediately? I mean how about `optimizer.target.cleargrads()` here.

File: espnet/asr/chainer_backend/asr.py [hunk 5]
```diff
@@ -154,8 +166,9 @@ def update_core(self):
             # update
             if math.isnan(grad_norm):
                 logging.warning('grad norm is nan. Do not update model.')
-            else:
+            elif self.count % self.accum_grad == 0:
                 optimizer.update()
+                self.clean = True
```
Comment (ShigekiKarita) [inline espnet/asr/chainer_backend/asr.py]: ditto

**Comment by ShigekiKarita** [comment]
LGTM except for several comments above. For `--model_module` argument, I think following existing notation in the `transformation` module is the best.

**Comment by Fhrozen** [comment]
@ShigekiKarita san, thank you for your hard work and the comments.
I will submitting the PR with the modifications by tomorrow.

File: egs/wsj/asr2/run.sh [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+        --train-json ${feat_tr_dir}/data.json \
+        --valid-json ${feat_dt_dir}/data.json \
+        --elayers ${elayers} \
+        --eunits ${eunits} \
+        --subsample ${subsample} \
+        --dlayers ${dlayers} \
+        --dunits ${dunits} \
+        --aheads ${aheads} \
+        --adim ${adim} \
+        --mtlalpha ${mtlalpha} \
```
Comment (sw005320) [inline egs/wsj/asr2/run.sh]: We can keep this, but please add some WARNING in the python program that this is not implemented yet.

File: egs/wsj/asr2/run.sh [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+        ngpu=0
+
+        ${decode_cmd} JOB=1:${nj} ${expdir}/${decode_dir}/log/decode.JOB.log \
+            asr_recog.py \
+            --ngpu ${ngpu} \
+            --backend ${backend} \
+            --recog-json ${feat_recog_dir}/split${nj}utt/data.JOB.json \
+            --result-label ${expdir}/${decode_dir}/data.JOB.json \
+            --model ${expdir}/results/${recog_model}  \
+            --beam-size ${beam_size} \
```
Comment (sw005320) [inline egs/wsj/asr2/run.sh]: We can keep this, but please add some WARNING in the python program that this is not implemented yet.

File: egs/wsj/asr2/run.sh [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+            --penalty ${penalty} \
+            --maxlenratio ${maxlenratio} \
+            --minlenratio ${minlenratio} \
+            --ctc-weight ${ctc_weight} \
```
Comment (sw005320) [inline egs/wsj/asr2/run.sh]: We can keep this, but please add some WARNING in the python program that this is not implemented yet.

File: egs/wsj/asr2/run.sh [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+            --lm-weight ${lm_weight} \
```
Comment (sw005320) [inline egs/wsj/asr2/run.sh]: We can keep this, but please add some WARNING in the python program that this is not implemented yet.

File: espnet/nets/chainer_backend/asr_transformer/attention.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
+# encoding: utf-8
```
Comment (sw005320) [inline espnet/nets/chainer_backend/asr_transformer/attention.py]: This function must be at `espnet/espnet/nets/chainer_backend/`
How about `espnet/espnet/nets/chainer_backend/attention_transformer.py`?

File: espnet/nets/chainer_backend/asr_transformer/__init__.py [hunk 11]
```diff
@@ -0,0 +1,4 @@
+from espnet.nets.chainer_backend.asr_transformer.e2e import E2E  # NOQA
```
Comment (sw005320) [inline espnet/nets/chainer_backend/asr_transformer/__init__.py]: I'm not a big fan of making `asr_transformer` directory. I think `asr` directory should be OK.

File: espnet/nets/chainer_backend/asr_transformer/decoders.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
+# encoding: utf-8
```
Comment (sw005320) [inline espnet/nets/chainer_backend/asr_transformer/decoders.py]: This function must be at `espnet/espnet/nets/chainer_backend/`
How about `espnet/espnet/nets/chainer_backend/decoders_transformer.py`?

File: espnet/nets/chainer_backend/asr_transformer/e2e.py [hunk 13]
```diff
@@ -0,0 +1,328 @@
+# encoding: utf-8
```
Comment (sw005320) [inline espnet/nets/chainer_backend/asr_transformer/e2e.py]: This function must be at `espnet/espnet/nets/chainer_backend/`.
How about `espnet/espnet/nets/chainer_backend/e2e_asr_transformer.py`?

File: espnet/nets/chainer_backend/asr_transformer/encoders.py [hunk 14]
```diff
@@ -0,0 +1,118 @@
+# encoding: utf-8
```
Comment (sw005320) [inline espnet/nets/chainer_backend/asr_transformer/encoders.py]: How about `espnet/espnet/nets/chainer_backend/encoders_transformer.py`?

File: espnet/nets/chainer_backend/asr_transformer/net_utils.py [hunk 15]
```diff
@@ -0,0 +1,171 @@
+# encoding: utf-8
```
Comment (sw005320) [inline espnet/nets/chainer_backend/asr_transformer/net_utils.py]: Add this one at `espnet/espnet/nets/chainer_backend/nets_utils_transformer.py`

File: egs/wsj/asr2/run.sh (hunk 6 above)
Comment (Fhrozen) [reply egs/wsj/asr2/run.sh]: The code internally raise an NotImplemented Flag:

https://github.com/Fhrozen/espnet/blob/0b77d8d39b850b4a6373ff0feae6c324a76e28f4/espnet/nets/chainer_backend/asr_transformer/e2e.py#L106-L107

Should I change it for a warning?

File: egs/wsj/asr2/run.sh (hunk 7 above)
Comment (Fhrozen) [reply egs/wsj/asr2/run.sh]: Same here:

https://github.com/Fhrozen/espnet/blob/0b77d8d39b850b4a6373ff0feae6c324a76e28f4/espnet/nets/chainer_backend/asr_transformer/e2e.py#L306-L307

Using raiseError

**Comment by Fhrozen** [comment]
Currently testing the new code:
Some points to resolve (or modify):
- [ ] Add results
- [ ] Add transformer test for chainer
- [ ] Dynamic loading. I am honestly not sure about this part because it will require to modify all recipes, Currently, the modifications just affect the transformer recipe. Maybe this could target in different PR.

File: egs/wsj/asr2/run.sh (hunk 6 above)
Comment (sw005320) [reply egs/wsj/asr2/run.sh]: OK, thanks

**Comment by ShigekiKarita** [comment]
> Dynamic loading
OK. Maybe I will handle this in the pytorch PR later.

**Comment by Fhrozen** [comment]
I just finished an additional commit:
- Trained and decoded for the large model and added the results.
- Added a test for chainer transfomer (train linear and decode). I am not sure about the other tests.
- resolved the cleargrad part. It seems that when training with single GPU and the grads are not cleared at the beginning, then the model does not update (grad_norm is nan). So I just added a cleargrad if count == 1, For multigpu the training seems to have no problem.

Let me know if there is any additional point to resolve or fix.

**Comment by ShigekiKarita** [comment]
LGTM :checkered_flag:

**Comment by sw005320** [comment]
Thanks @Fhrozen!!!
@ShigekiKarita, please follow up it!
